suite/tests/client-interface/mangle_suspend.dll.c

```diff
@@ -0,0 +1,190 @@
+ *   used to endorse or promote products derived from this software without
+ *   specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL GOOGLE, INC. OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+/* Test xl8 pc of rip-rel instruction (xref #3307) caused by
+ * asynch interrupt.
+ */
+
+#include "dr_api.h"
+#include "drmgr.h"
+#include "client_tools.h"
+#include "mangle_suspend-shared.h"
+#include <string.h> /* memset */
+
+#define MAX_RESUME_COUNT 10
+
+#define CHECK(x, msg)                                                                \
+    do {                                                                             \
+        if (!(x)) {                                                                  \
+            dr_fprintf(STDERR, "CHECK failed %s:%d: %s\n", __FILE__, __LINE__, msg); \
+            dr_abort();                                                              \
+        }                                                                            \
+    } while (0);
+
+extern void
+test_asm();
+
+static byte *add_instr_pc = NULL;
+
+static dr_emit_flags_t
+event_app_analysis(void *drcontext, void *tag, instrlist_t *bb, bool for_trace,
+                   bool translating, OUT void **user_data)
+{
+    instr_t *inst;
+    bool prev_was_mov_const = false;
+    ptr_int_t val1, val2;
+    *user_data = NULL;
+
+    if (translating)
+        return DR_EMIT_DEFAULT;
+
+#ifdef X86_64
+    /* Look for duplicate mov immediates telling us which subaction we're in */
+    for (inst = instrlist_first_app(bb); inst != NULL; inst = instr_get_next_app(inst)) {
+        if (instr_is_mov_constant(inst, prev_was_mov_const ? &val2 : &val1)) {
+            if (prev_was_mov_const && val1 == val2 &&
+                val1 != 0 && /* rule out xor w/ self */
+                opnd_is_reg(instr_get_dst(inst, 0)) &&
+                opnd_get_reg(instr_get_dst(inst, 0)) == SUSPEND_TEST_REG) {
+                *user_data = (void *)val1;
+                instrlist_meta_postinsert(bb, inst, INSTR_CREATE_label(drcontext));
+            } else
+                prev_was_mov_const = true;
+        } else
+            prev_was_mov_const = false;
+    }
+#else
+        /* XXX i#3307: port to ARM if possible. */
```

Comment:
This means you are not closing the issue, then?  Please add a comment to the end of the issue once this PR is in about what remains to be done, and maybe update the title, and update the labels (right now the labels show it's x86-only).

Reply:
Changed comment to i#3329